content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+4. Csmith uses a set of *saftey checks* to make sure the newly generated bit of code is well formed, following design goal #1. If a safety check fails, the changes are undone and we try again.
+
+TODO: Provide an example.
+
+## Safety Mechanisms
+
+A program can crash for various reasons. It might be because of compiler failures, but more often, due to unsafe programs. Csmith wants to avoid these pitfalls with proper safety mechanisms.
+
+**Discussion:** Think about what safety issues can be caused by the randomly generated programs.
+
+A summary of all mechanisms shown below. We will provide detailed explanations for it.
+
+<img src="safety_mechanism.jpg" style="width: 70%">
+
+### Integer Safety
+
+The safety problem of integers comes from undefined behaviors (UB) such as signed overflow:
+
+```c++
+int signedOverflow(int x) {
+    return x+1 > x; // either true or UB due to signed overflow
+}
+```
+
+and shift-past-bitwidth:
+
+```c++
+int shiftPastBitwidth() {
+    return 1 << 32; // UB when evaluated on 32 bit platform 
+}
+```
+
+Csmith generates wrapper functions for arithmetic operators whose operands might overflow according to the compiler standard, though there are tricky UBs not defined and the authors have to figure out themselves (such as `INT_MIN % -1`).
+
+### Type Safety
+
+The tricky aspect of C's type safety is qualifier safety. Modifying constant-qualified or volatile-qualified objects through nonconstant references is [undefined behavior](<https://wiki.sei.cmu.edu/confluence/display/c/EXP40-C.+Do+not+modify+constant+objects>).
+
+```c++
+const int **ipp; // the value pointer to shall not be modified
+int *ip;
+const int i = 42;
+ 
+void constViolation(void) {
+  ipp = &ip; // UB
+  *ipp = &i; // Valid
+  *ip = 0;   // Modifies constant i (was 42)
+}
+```
+
+Csmith ensures type safety with type checks.
+
+### Pointer safety
+
+The first kind of pointer safety problem is null pointer dereference.
+
+```c++
+int a = 10;
+void nullDereference(int *p){
+    *pi = a; // cause execption if p is NULL
+}
+```
+
+ This can be easily avoided by dynamic checks.
+
+```c++
+int a = 10;
+void safeDereference(int *p){
+    if( p != NULL){
+    	*p = a; 
+    }
+}
+```
+
+However, there is no reliable method to identify an invalid pointer that points to a function-scoped variable.
+
+```c++
+void invalidDereference(int *p){
+    int a = 10;
+    if( p != NULL){
+    	*pi = a; 
+    }
+}
+// outside this function, we cannot dereference or compare p with other pointer 
+// before it becomes valid again!
+```
+
+**Question:** What could be the solution here? 
+
+One way is to force a pointer, which points to a function-scoped variable, to never outlive the function; however, this is too restrictive. Csmith instead chooses to do pointer analysis that is flow sensitive, field sensitive, context sensitive, path insensitive, and array-element insensitive. 
```

Comment:
A little more detail would be useful: what does it *do* with that pointer analysis? That is, just having the aliasing information doesn't do anything by itself—you have to *use* the aliasing information to rule out some kind of program behavior.